ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/plugin/ExecutionEnvironmentSection.java

```diff
@@ -283,17 +300,62 @@ private void handleRemove() {
 	}
 
 	private void handleAdd() {
-		ElementListSelectionDialog dialog = new ElementListSelectionDialog(PDEPlugin.getActiveWorkbenchShell(), new EELabelProvider());
+		ElementListSelectionDialog dialog = showDialog(false);
+		if (dialog.open() == Window.OK) {
+			addExecutionEnvironments(dialog.getResult());
+		}
+	}
+	private void handleUpgrade() {
+
+		RequiredExecutionEnvironmentHeader header = getHeader();
+		IExecutionEnvironment[] executionEnvironments = JavaRuntime.getExecutionEnvironmentsManager()
+				.getExecutionEnvironments();
+		if (executionEnvironments.length > 0) {
+			ElementListSelectionDialog dialog = showDialog(true);
+			if (dialog.open() == Window.OK) {
+
+				List<String> addedExecutionEnvironments = header.getEnvironments();
+				// Remove existing EE ( Upgrade is allowed only when single Java
+				// EE is present in the header)
+				getHeader().removeExecutionEnvironment(addedExecutionEnvironments.get(0));
+
+				// add selected Java EE to the header
+				addExecutionEnvironments(dialog.getResult());
+
+				final IProject project = getPage().getPDEEditor().getCommonProject();
+				updateClasspathSettings(project);
+			}
+		}
+	}
+
+	private ElementListSelectionDialog showDialog(boolean isUpgrade) {
+		ElementListSelectionDialog dialog = new ElementListSelectionDialog(PDEPlugin.getActiveWorkbenchShell(),
+				new EELabelProvider()) {
+			@Override
+			protected FilteredList createFilteredList(Composite parent) {
+				FilteredList filteredList = super.createFilteredList(parent);
+				filteredList.setComparator(Comparator.comparing(ee -> new VersionComparator(getVersionString(ee))));
```

Comment:
When https://github.com/eclipse-pde/eclipse.pde/pull/1324 is completed, this can just be
```suggestion
				filteredList.setComparator(VMUtil.ASCENDING_EE_JAVA_VERSION.reversed());
```